File: src/main/java/cz/cvut/fel/plichjan/distmesh/FreeFemMeshReader.java
package cz.cvut.fel.plichjan.distmesh;

import cz.cvut.fel.plichjan.distmesh.result.Mesh;

import java.io.BufferedReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * http://www.ann.jussieu.fr/~lehyaric/ffcs/doc/ff/ffdocsu17.html
 */
public class FreeFemMeshReader extends TriangleFormatReader {
    private int nVertex;
    private int nTriangles;
    private List<int[]> triangles;
    private List<double[]> nodes;

    public FreeFemMeshReader(String name) throws IOException {
        BufferedReader msh = initTokenizer(name + ".msh");
        parseHeader();
        parseVertices();
        parseTriangles();
        msh.close();
    }

    public Mesh getMesh() {
        final Mesh mesh = new Mesh();
        mesh.setP(nodes.toArray(new double[nodes.size()][]));
        mesh.setT(triangles.toArray(new int[triangles.size()][]));
        return mesh;
    }

    private void parseHeader() throws IOException {
        // parse header
        nVertex = readInt();
        nTriangles = readInt();
        readInt(); // the number of edges on boundary
    }

    private void parseTriangles() throws IOException {
        triangles = new ArrayList<int[]>(nTriangles);

        for (int i = 0; i < nTriangles; i++) {
            triangles.add(readIntArr(3));
            readInt(); // triangle label
//            readEOL();
        }
    }

    private void parseVertices() throws IOException {
        nodes = new ArrayList<double[]>(nVertex);

        for (int i = 0; i < nVertex; i++) {
            nodes.add(readDoubles(2));
            readInt(); // vertex label
//            readEOL();
        }
    }

}

File: src/main/kotlin/cz/cvut/fel/plichjan/distmesh/FreeFemMeshReader.kt
package cz.cvut.fel.plichjan.distmesh

import cz.cvut.fel.plichjan.distmesh.result.Mesh
import java.io.IOException
import java.util.ArrayList

class FreeFemMeshReader @Throws(IOException::class) constructor(name: String) : TriangleFormatReader() {
    private var nVertex = 0
    private var nTriangles = 0
    private var triangles: MutableList<IntArray>? = null
    private var nodes: MutableList<DoubleArray>? = null

    init {
        val msh = initTokenizer("$name.msh")
        parseHeader()
        parseVertices()
        parseTriangles()
        msh.close()
    }

    val mesh: Mesh
        get() {
            val mesh = Mesh()
            mesh.p = nodes!!.toTypedArray()
            mesh.t = triangles!!.toTypedArray()
            return mesh
        }

    @Throws(IOException::class)
    private fun parseHeader() {
        nVertex = readInt()
        nTriangles = readInt()
        readInt() // number of edges on boundary
    }

    @Throws(IOException::class)
    private fun parseTriangles() {
        triangles = ArrayList(nTriangles)
        for (i in 0 until nTriangles) {
            triangles!!.add(readIntArr(3))
            readInt() // triangle label
        }
    }

    @Throws(IOException::class)
    private fun parseVertices() {
        nodes = ArrayList(nVertex)
        for (i in 0 until nVertex) {
            nodes!!.add(readDoubles(2))
            readInt() // vertex label
        }
    }
}
